meetings/working-groups/field-keyword/FK-2024-08-07 Nullability analysis with the `field` keyword.md

```diff
@@ -0,0 +1,373 @@
+# Nullability analysis with the `field` keyword
+
+Table of contents:
+
+- [Motivation](#motivation)
+- [Introduction](#introduction)
+- [Constructor initialization](#constructor-initialization)
+- [Field-targeted nullability attributes](#field-targeted-nullability-attributes)
+- [Property-targeted nullability attributes](#property-targeted-nullability-attributes)
+- [Null-resilient getters](#null-resilient-getters)
+- [Alternatives](#alternatives)
+
+## Motivation
+
+Some vastly common patterns which are served by the `field` feature are lazy initialization, and defaulting. When a field-backed property is non-nullable, it's required to be initialized during construction. This flies in the face of these kinds of properties. They are well-formed. Warnings requiring constructor initialization would be noisy and unwelcome in these properties and in many similar forms.
+
+```cs
+public List<int> LazyProp => field ??= new();
+
+public string DefaultingProp { get => field ?? GetDefault(); private set; }
+```
+
+These scenarios are made possible using the existing nullability attributes, like `[MaybeNull]`, just as they would be for manually-declared fields.
+
+We'll walk first through enabling the user to provide these nullability attributes explicitly for the backing field, and then through how the compiler builds on this capability to apply these attributes automatically in a simple and intuitive fashion.
+
+## Introduction
+
+This proposal brings together a few building blocks:
+
+1. Properties can be considered initialized to non-null in the constructor by calling the setter, just as with `required` and normal flow analysis today.
+
+1. `field`-targeted nullability attributes work just as with manually declared fields. (E.g. `[field: MaybeNull]`)
+
+1. Null-resilient getters, a simple concept that allows maintaining a property's contract of not returning null, while flexibly allowing idiomatic patterns that use a nullable backing field. 
+
+The following sections will walk through how these concepts play out.
+
+## Constructor initialization
+
+In C# today, a non-nullable auto property is required to be initialized in the constructor. We will preserve this requirement for all field-backed properties. This requirement is in the form of the existing warning "CS8618: Non-nullable property 'Prop' must contain a non-null value when exiting constructor."
```

Comment:
```suggestion
In C# today, a non-nullable auto property that is not marked `required` is forced to be initialized in the constructor. We will preserve this enforcement for all field-backed properties. This enforcement is in the form of the existing warning "CS8618: Non-nullable property 'Prop' must contain a non-null value when exiting constructor."
```

Reply:
At this level of detail, I should probably also add "that has no property initializer and..."

Reply:
```suggestion
In C# today, a non-nullable auto property which has no property initializer and is not marked `required` is forced to be initialized in the constructor. We will preserve this enforcement for all field-backed properties. This enforcement is in the form of the existing warning "CS8618: Non-nullable property 'Prop' must contain a non-null value when exiting constructor."
```